book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+
+### The Widest Type: `unknown`
+
+TypeScript's widest type is `unknown`. It represents something that we don't know what it is.
+
+If you imagine a scale whether the widest types are at the top and the narrowest types are at the bottom, `unknown` is at the top. All other types like strings, numbers, booleans, null, undefined, and their respective literals are assignable to `unknown`, as seen in its assignability chart:
+
+<img src="https://res.cloudinary.com/total-typescript/image/upload/v1706814781/065-introduction-to-unknown.explainer_ohm9pd.png">
+
+Consider this example function `fn` that takes in an `input` parameter of type `unknown`:
+
+```ts twoslash
+const fn = (input: unknown) => {};
+
+// Anything is assignable to unknown!
+fn("hello");
+fn(42);
+fn(true);
+fn({});
+fn([]);
+fn(() => {});
+```
+
+All of the above function calls are valid because `unknown` is assignable to any other type
+
+The `unknown` type is the preferred choice when you want to represent something that's truly unknown in JavaScript. For example, it is extremely useful when you have things coming into your application from outside sources, like input from a form or a call to a webhook.
+
+#### What's the Difference Between `unknown` and `any`?
+
+You might be wondering what the difference is between `unknown` and `any`. They're both wide types, but there's a key difference.
+
+`any` doesn't really fit into our definition of 'wide' and 'narrow' types. It breaks the type system. It's not really a type at all - it's a way of opting out of TypeScript's type checking.
+
+`any` can be assigned to anything, and anything can be assigned to `any`. `any` is both narrower and wider than every other type.
+
+`unknown`, on the other hand, is part of TypeScript's type system. It's wider than every other type, so it can't be assigned to anything.
+
+```ts twoslash
+// @errors: 18046
+const handleWebhookInput = (input: unknown) => {
+  input.toUppercase();
+};
+
+const handleWebhookInputWithAny = (input: any) => {
+  // no error
+  input.toUppercase();
+};
+```
+
+This means that `unknown` is a safe type, but `any` is not. `unknown` means "I don't know what this is", while `any` means "I don't care what this is".
+
+### The Narrowest Type: `never`
+
+If `unknown` is the widest type in TypeScript, `never` is the narrowest.
+
+`never` represents something that will _never_ happen. It's the very bottom of the type hierarchy.
+
+You'll rarely use a `never` type annotation yourself. Instead, it'll pop up in error messages and hovers - often when narrowing.
+
+But first, let's look at a simple example of a `never` type:
+
+#### `never` vs `void`
+
+Let's consider a function that never returns anything:
+
+```typescript
+const getNever = () => {
+  // This function never returns!
+};
+```
+
+When hovering this function, TypeScript will infer that it returns `void`, indicating that it essentially returns nothing.
```

Comment:
`void` means that it returns anything, not nothing. In general, I just find `void` to be obscure in TS and often `unknown` might be a better option. You can accidentally run into issues with `void` cause it's "anything" but yet TS sometimes assumes it's `undefined`(-like?). It's very confusing :v